File: reader.py
"""
reader.py — fetch messages from public Telegram channels via Telethon (MTProto).
Uses a StringSession so no interactive login is needed in CI.
"""

from __future__ import annotations

import logging
from datetime import datetime, timedelta, timezone
from dataclasses import dataclass

from telethon import TelegramClient
from telethon.sessions import StringSession
from telethon.tl.types import Message

from config import TG_API_ID, TG_API_HASH, TG_SESSION_STRING, LOOKBACK_HOURS, MAX_MSGS_PER_CH

log = logging.getLogger(__name__)


@dataclass
class NewsItem:
    channel: str
    message_id: int
    date: datetime
    text: str
    url: str


def _build_url(channel: str, msg_id: int) -> str:
    username = channel.lstrip("@")
    return f"https://t.me/{username}/{msg_id}"


async def fetch_recent_messages(channel: str) -> list[NewsItem]:
    """Return up to MAX_MSGS_PER_CH messages from *channel* posted in the last LOOKBACK_HOURS."""
    cutoff = datetime.now(timezone.utc) - timedelta(hours=LOOKBACK_HOURS)
    items: list[NewsItem] = []

    async with TelegramClient(
        StringSession(TG_SESSION_STRING), TG_API_ID, TG_API_HASH
    ) as client:
        async for msg in client.iter_messages(channel, limit=MAX_MSGS_PER_CH):
            if not isinstance(msg, Message):
                continue
            if msg.date < cutoff:
                break                           # messages are newest-first; stop early
            text = (msg.text or "").strip()
            if not text or len(text) < 20:      # skip stickers / very short service msgs
                continue
            items.append(
                NewsItem(
                    channel=channel,
                    message_id=msg.id,
                    date=msg.date,
                    text=text,
                    url=_build_url(channel, msg.id),
                )
            )
            if len(items) >= MAX_MSGS_PER_CH:
                break

    log.debug(f"{channel}: {len(items)} usable messages since {cutoff.isoformat()}")
    return items

File: reader.js
/**
 * reader.js — fetch recent messages from public Telegram channels via GramJS (MTProto).
 * Uses a StringSession so no interactive login is needed in CI.
 */

import { TelegramClient } from "telegram";
import { StringSession }  from "telegram/sessions/index.js";
import {
  TG_API_ID,
  TG_API_HASH,
  TG_SESSION_STRING,
  LOOKBACK_HOURS,
  MAX_MSGS_PER_CH,
} from "./config.js";

const CUTOFF_MS = LOOKBACK_HOURS * 60 * 60 * 1000;

function buildUrl(channel, msgId) {
  const username = channel.replace(/^@/, "");
  return `https://t.me/${username}/${msgId}`;
}

let _client = null;

/**
 * Initialize the Telegram client once.
 */
export async function initReader() {
  if (_client) return;
  const session = new StringSession(TG_SESSION_STRING);
  _client = new TelegramClient(session, TG_API_ID, TG_API_HASH, {
    connectionRetries: 3,
  });
  await _client.connect();
}

/**
 * Destroy the Telegram client to kill background update loops.
 */
export async function destroyReader() {
  if (_client) {
    if (typeof _client.destroy === "function") {
      await _client.destroy();
    } else {
      await _client.disconnect();
    }
    _client = null;
  }
}

/**
 * @returns {Promise<Array<{ channel, messageId, date, text, url }>>}
 */
export async function fetchRecentMessages(channel) {
  if (!_client) throw new Error("Reader not initialized. Call initReader() first.");

  const cutoff = Date.now() - CUTOFF_MS;
  const items  = [];

  const messages = await _client.getMessages(channel, { limit: MAX_MSGS_PER_CH });

  for (const msg of messages) {
    const ts = msg.date * 1000; // GramJS uses Unix seconds
    if (ts < cutoff) continue;

    const text = (msg.message || "").trim();
    if (!text || text.length < 20) continue; // skip stickers / empty

    items.push({
      channel,
      messageId: msg.id,
      date:      new Date(ts).toISOString(),
      text,
      url:       buildUrl(channel, msg.id),
    });
  }

  console.log(`[reader] ${channel}: ${items.length} messages fetched`);
  return items;
}
